pkg/solana/logpoller/log_poller.go

```diff
@@ -0,0 +1,316 @@
+	if err != nil {
+		return fmt.Errorf("failed to load filters: %w", err)
+	}
+
+	lp.filtersMutex.Lock()
+	defer lp.filtersMutex.Unlock()
+
+	filter, ok := lp.filtersByName[name]
+	if !ok {
+		lp.lggr.Warnw("Filter not found in filtersByName", "name", name)
+		return nil
+	}
+
+	if err := lp.orm.MarkFilterDeleted(ctx, filter.ID); err != nil {
+		return fmt.Errorf("failed to mark filter deleted: %w", err)
+	}
+
+	delete(lp.filtersByName, filter.Name)
+	filtersByEventSig, ok := lp.filtersByAddress[filter.Address]
+	if !ok {
+		lp.lggr.Warnw("Filter not found in filtersByAddress", "name", name, "address", filter.Address)
+		return nil
+	}
+
+	filtersByEventSig[filter.EventSig], ok = removeFilterFromSlice(filtersByEventSig[filter.EventSig], filter)
+	if !ok {
+		lp.lggr.Warnw("Filter not found in filtersByEventSig", "name", name, "address", filter.Address)
+	}
+
+	if len(filtersByEventSig[filter.EventSig]) == 0 {
+		delete(filtersByEventSig, filter.EventSig)
+	}
+
+	if len(lp.filtersByAddress[filter.Address]) == 0 {
+		delete(lp.filtersByAddress, filter.Address)
+	}
+
+	// remove or ensure that filters was not present in the slice to backfill
+	lp.filtersToBackfill, _ = removeFilterFromSlice(lp.filtersToBackfill, filter)
+	lp.filtersToDelete = append(lp.filtersToDelete, filter)
+	return nil
+}
+
+func (lp *LogPoller) loadFilters(ctx context.Context) error {
+	if lp.loadedFilters.Load() {
+		return nil
+	}
+
+	lp.lggr.Debugw("Loading filters from db")
+	lp.filtersMutex.Lock()
+	defer lp.filtersMutex.Unlock()
+	// reset filters' indexes to ensure we do not have partial data from the previous run
+	lp.filtersByAddress = make(map[PublicKey]map[EventSignature][]Filter)
+	lp.filtersByName = make(map[string]Filter)
+	lp.filtersToBackfill = nil
+	lp.filtersToDelete = nil
+
+	ctx, cancel := lp.chStop.Ctx(ctx)
+	defer cancel()
+	filters, err := lp.orm.SelectFilters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to select filters from db: %w", err)
+	}
+
+	for _, filter := range filters {
+		if filter.IsDeleted {
+			lp.filtersToDelete = append(lp.filtersToDelete, filter)
+			continue
+		}
+
+		if _, ok := lp.filtersByName[filter.Name]; ok {
+			errMsg := fmt.Sprintf("invariant violation while loading from db: expected filters to have unique name: %s ", filter.Name)
+			lp.lggr.Critical(errMsg)
+			return errors.New(errMsg)
+		}
+
+		lp.filtersByName[filter.Name] = filter
+		filtersByEventSig, ok := lp.filtersByAddress[filter.Address]
+		if !ok {
+			filtersByEventSig = make(map[EventSignature][]Filter)
+			lp.filtersByAddress[filter.Address] = filtersByEventSig
+		}
+
+		filtersByEventSig[filter.EventSig] = append(filtersByEventSig[filter.EventSig], filter)
+		lp.filtersToBackfill = append(lp.filtersToBackfill, filter)
+	}
+
+	lp.loadedFilters.Store(true)
+	return nil
+}
+
+func removeFilterFromSlice(filters []Filter, filter Filter) ([]Filter, bool) {
+	index := slices.IndexFunc(filters, func(item Filter) bool {
+		return item.ID == filter.ID
+	})
```

Comment:
We could make the type of filtersByAddress be:
```
filtersByAddress map[PublicKey]map[EventSignature]map[ID]Filter
```
instead of:
```
filtersByAddress map[PublicKey]map[EventSignature][]Filter
```
Seems like that would be more efficient and simplify this removal function. What do you think?

Reply:
Oh yeah, we'd have to do the same with filtersToBackfill:
```
--- filtersToBackfill []Filter
+++ filtersToBackfill map[int64]Filter
```

Reply:
It's easier to delete from `filtersByAddress map[PublicKey]map[EventSignature]map[int64]Filter`, but more difficult to write as we need additional nil check.

I guess I was just lazy. Switched to the map implementation.